### PR TITLE
Fix/2031  limit number of sharespaces

### PR DESCRIPTION
**Pull request by inkhey**
related to #2031 

## Checkpoints

The goal of this section is to help code reviewer to do required checks on this pull request. Please don't edit this list.
If one or more of these checkpoints are out of scope, please write below why they are.

- [x] Code is clear enough
- [x] If there are FIXME in the code, then there are associated issues and issue is referenced in the FIXME
- [x] If there are TODO, NOTE or HACK in code, date and developer initials are present
- [x] Automated tests have been written (covering feature or non regression), or *at least* an issue has been created for test implementation

### CODE REVIEW

File: backend/tracim_backend/lib/agenda/agenda.py [hunk 1]
```diff
@@ -193,7 +193,7 @@ def get_user_agendas(
             workspace_api = WorkspaceApi(
                 current_user=self._user, session=self._session, config=self._config
             )
-            workspaces = workspace_api.get_all_for_user(user)
+            workspaces = workspace_api.get_all_for_user(user, with_role=True, owned=False)
```
Comment (buxx) [inline backend/tracim_backend/lib/agenda/agenda.py]: Why exclude owned workspaces here ?

Comment (inkhey) [reply backend/tracim_backend/lib/agenda/agenda.py]: This code excludes only workspace where user is owner and has no role.
We do not expect user without any role to access to agenda feature. This change is just to obtain same behavior as before.

Comment (buxx) [reply backend/tracim_backend/lib/agenda/agenda.py]: Okay I got it. `with_role` and `owned` permit to include them. If both are set to False, list is always empty.
Can you rename into `include_with_role` and `include_owned` and add a docstring to `get_all_for_user` please ?

File: backend/tracim_backend/lib/core/workspace.py [hunk 2]
```diff
@@ -83,6 +84,11 @@ def create_workspace(
         agenda_enabled: bool = True,
         save_now: bool = False,
     ) -> Workspace:
+        from tracim_backend.lib.core.user import UserApi
```
Comment (buxx) [inline backend/tracim_backend/lib/core/workspace.py]: When import is not at top of file for a specific reason, it is preferable to indicate why in a note.

File: backend/tracim_backend/lib/core/workspace.py [hunk 3]
```diff
@@ -178,20 +184,35 @@ def get_one_for_current_user(self, id):
     def get_all(self):
         return self._base_query().all()
 
-    def get_all_for_user(self, user: User, ignored_ids=None):
-        workspaces = []
+    def _get_workspace_owned_by_user(self, user_id: int) -> typing.List[Workspace]:
```
Comment (buxx) [inline backend/tracim_backend/lib/core/workspace.py]: Add a `s` to `workspace` ?

File: backend/tracim_backend/lib/core/workspace.py [hunk 4]
```diff
@@ -178,20 +184,35 @@ def get_one_for_current_user(self, id):
...
+        return self._base_query_without_roles().filter(Workspace.owner_id == user_id).all()
+
+    def _get_workspace_ids(self, workspaces: typing.List[Workspace]) -> typing.List[int]:
+        workspace_ids = []
+        for workspace in workspaces:
+            workspace_ids.append(workspace.workspace_id)
+        return workspace_ids
```
Comment (buxx) [inline backend/tracim_backend/lib/core/workspace.py]: I'm not sure specific method is useful for that. @lebouquetin What is your opinion about use short code like following instead a real function ?

```python
[w.id for w in workspaces]
```

File: backend/tracim_backend/lib/webdav/resources.py [hunk 5]
```diff
@@ -176,7 +179,14 @@ def createCollection(self, name: str):
...
         workspace_name = webdav_convert_file_name_to_bdd(name)
-        new_workspace = self.workspace_api.create_workspace(workspace_name)
+        try:
+            new_workspace = self.workspace_api.create_workspace(workspace_name)
+        except (
+            UserNotAllowedToCreateMoreWorkspace,
+            EmptyLabelNotAllowed,
+            WorkspaceLabelAlreadyUsed,
+        ):
+            raise DAVError(HTTP_FORBIDDEN)
```
Comment (buxx) [inline backend/tracim_backend/lib/webdav/resources.py]: `DAVError` class own `contextinfo` (`str`) parameter in constructor: we can use it to add some information about error (like `DAVError(HTTP_FORBIDDEN, contextinfo=str(exc)`).

Comment (inkhey) [reply backend/tracim_backend/lib/webdav/resources.py]: Added this to all webdav case Forbidden error where it's possible to have explicit error.

Comment (inkhey) [reply backend/tracim_backend/lib/webdav/resources.py]: This is a nice feature for debugging, but those informations are not provided by webdav client like nautilus (don't know if some client does give thoses informations).

File: backend/tracim_backend/lib/core/workspace.py (hunk 3 above)
Comment (inkhey) [reply backend/tracim_backend/lib/core/workspace.py]: done.

File: backend/tracim_backend/lib/core/workspace.py (hunk 2 above)
Comment (inkhey) [reply backend/tracim_backend/lib/core/workspace.py]: explained.

File: backend/tracim_backend/lib/agenda/agenda.py (hunk 1 above)
Comment (inkhey) [reply backend/tracim_backend/lib/agenda/agenda.py]: done.

File: backend/tracim_backend/lib/core/workspace.py (hunk 4 above)
Comment (lebouquetin) [reply backend/tracim_backend/lib/core/workspace.py]: @buxx @inkhey this way of writing code may be complicated to understand... when it is combined with conditionnal options and sub-inlining operations.

The example you write is ok: the code is clear and easy to read and understand

Comment (inkhey) [reply backend/tracim_backend/lib/core/workspace.py]: fixed.

**Pull request merged**